libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp

```diff
@@ -20,42 +20,68 @@ extern const AP_HAL::HAL& hal;
 // Control reset of yaw and magnetic field states
 void NavEKF2_core::controlMagYawReset()
 {
-    // Use a quaternion division to calcualte the delta quaternion between the rotation at the current and last time
-    Quaternion deltaQuat = stateStruct.quat / prevQuatMagReset;
-    prevQuatMagReset = stateStruct.quat;
+    // Quaternion and delta rotation vector that are re-used for different calculations
+    Vector3f deltaRotVecTemp;
+    Quaternion deltaQuatTemp;
 
-    // convert the quaternion to a rotation vector and find its length
-    Vector3f deltaRotVec;
-    deltaQuat.to_axis_angle(deltaRotVec);
+    bool flightResetAllowed = false;
+    bool initialResetAllowed = false;
+    if (!finalInflightYawInit) {
+        // Use a quaternion division to calculate the delta quaternion between the rotation at the current and last time
+        deltaQuatTemp = stateStruct.quat / prevQuatMagReset;
+        prevQuatMagReset = stateStruct.quat;
 
-    // check if the spin rate is OK - high spin rates can cause angular alignment errors
-    bool angRateOK = deltaRotVec.length() < 0.1745f;
+        // convert the quaternion to a rotation vector and find its length
+        deltaQuatTemp.to_axis_angle(deltaRotVecTemp);
 
-    // a flight reset is allowed if we are not spinning too fast and are in flight
-    bool flightResetAllowed = angRateOK && inFlight;
+        // check if the spin rate is OK - high spin rates can cause angular alignment errors
+        bool angRateOK = deltaRotVecTemp.length() < 0.1745f;
+
+        initialResetAllowed = angRateOK;
+        flightResetAllowed = angRateOK && !onGround;
+
+    }
 
     // check that we have reached a height where ground magnetic interference effects are insignificant
-    bool hgtCheckPassed = (stateStruct.position.z  - posDownAtTakeoff) < -5.0f;
+    // and can perform a final reset of the yaw and field states
+    bool finalResetRequest = false;
+    if (flightResetAllowed && !assume_zero_sideslip()) {
+        finalResetRequest = (stateStruct.position.z  - posDownAtTakeoff) < -5.0f;
+    }
+
+    // Check for bad yaw casued by ground based magnetic anomaly
+    bool interimResetRequest = false;
+    if (flightResetAllowed && !assume_zero_sideslip()) {
```

Comment:
Any reason for having the repeated if? Or was it not supposed to be the same if?